### PR TITLE
Support input aml model run locally

### PR DESCRIPTION
**Pull request by xiaoyu-work**
## Describe your changes

1. Add AML model download logic.
2. Since we have more use cases for AML client other than AMLSystem. Extract AML client creation logic out to a standalone logic.
3. Support input aml model run locally.

Notes:
Since OliveModel class is getting more and more AML model related attribute, we may need a more structured AML model class and local model class. 

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Format your code by running `pre-commit run --all-files`

### CODE REVIEW

File: examples/bert_hf_cpu_aml/bert_config.json [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
+        "resource_group": "<place_holder>",
+        "workspace_name": "<place_holder>"
+    },
     "input_model":{
         "type": "PyTorchModel",
         "config": {
             "name": "bert-hf",
             "version": "3",
             "model_storage_kind": "azureml",
+            "aml_storage_name": "bert_hf",
```
Comment (trajepl) [inline examples/bert_hf_cpu_aml/bert_config.json:14]: How about move aml_client info from engine entry to input model's property, Like:
image

File: olive/engine/engine.py [hunk 2]
```diff
@@ -207,6 +208,7 @@ def run(
         self,
         input_model: OliveModel,
         packaging_config: Optional[PackagingConfig] = None,
+        aml_client: Optional[AzureMLClient] = None,
```
Comment (guotuofeng) [inline olive/engine/engine.py:211]: do we really need to make the aml_client concept known to engine? 

Ideally, only model.py should download the model when the model is loaded.

[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply olive/engine/engine.py:211]: Yeah there are 2 options here: either put aml_client to model or engine. For me, Engine owns this makes more sense than model. Another use case for Engine will be that Engine will be responsible for packaging output models to AML in the near future. Then Engine will still need AML client.

Comment (xiaoyu-work) [reply olive/engine/engine.py:211]: Another use case for aml_client for Engine will be AML packaging. Engine will need to have aml_client to package out models to AML with AML_client. This part of logic is not sent as a PR for now but will be needed in the near future.

**Review by devang-ml — COMMENTED** [review]
We need to support packaging output models from Footprints to AML.

File: olive/engine/engine.py (hunk 2 above)
Comment (guotuofeng) [reply olive/engine/engine.py:211]: could we make the aml_client part of the packaging logic instead of engine?

File: olive/workflows/run/config.py [hunk 3]
```diff
@@ -136,6 +138,10 @@ def _resolve_config_str(v, values, alias, component_name="systems"):
     components = values[component_name] or {}
     if sub_component not in components:
         raise ValueError(f"{alias} {sub_component} not found in systems")
+    if component_name == "systems" and components[sub_component].type == "AzureML":
```
Comment (jambayk) [inline olive/workflows/run/config.py:141]: can you move this auto insert to `_resolve_system` at line `149`? 

I think it can be something like 
```
def _resolve_system(v, values, system_alias, component_name="systems"):
    v = _resolve_config_str(v, values, system_alias, component_name=component_name) 
    v[system_alias] = validate_config(v[system_alias], SystemConfig)
    if v[system_alias].type == "AzureML":
         .... auto insert code .... 
``` 

Otherwise, it will only be auto inserted if the system is referenced by name. We also accept system config dicts directly.

Comment (xiaoyu-work) [reply olive/workflows/run/config.py:141]: Sure will do. This default `component_name="systems"` confused me a lot. If there is no specific use case here, I'll remove this.

Comment (jambayk) [reply olive/workflows/run/config.py:141]: Yes, it will always be "systems" for `_resolve_system`, so it is redundant. 

`_resolve_config_str` also has direct references to `system` which should actually be `component_name`. Looks like the PR that generalized the `resolve_config_str` code missed them.

Comment (xiaoyu-work) [reply olive/workflows/run/config.py]: Yeah I mean the default value in `_resolve_config_str` for system. We also use it for data_config right now. I've removed all system keyword.

Comment (jambayk) [reply olive/workflows/run/config.py]: Thanks! it looks good to me now

**Comment by jambayk** [comment]
One of the ci jobs cancelled by itself. I think you will need to push an empty commit to retrigger the ci pipelines.

File: olive/azureml/azureml_client.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                raise ValueError(f"aml_config_path {v} does not exist")
+            if not Path(v).is_file():
+                raise ValueError(f"aml_config_path {v} is not a file")
+        return v
+
+    def create_client(self):
+        return AzureMLClient(**self.dict())
+
+
+class AzureMLClient(object):
```
Comment (guotuofeng) [inline olive/azureml/azureml_client.py:45]: do we really need AzureMLClien? or we can directly return MLClient in create_client?

File: olive/engine/engine.py [hunk 5]
```diff
@@ -668,9 +676,15 @@ def _run_passes(
                 model.model_storage_kind == ModelStorageKind.AzureMLModel
                 and not self.host_for_pass(pass_id).system_type == SystemType.AzureML
             ):
-                error_msg = "Azure ML model only supports AzureMLSystem for Olive Pass"
-                logger.error(error_msg)
-                raise Exception(error_msg)
+                model_download_path = self._model_cache_path / "azureml_input_model"
+                model_path = model.download_from_azureml(self.azureml_client, model_download_path)
```
Comment (guotuofeng) [inline olive/engine/engine.py:680]: >model [](http://example.com/codeflow?start=29&length=5)

what about the input model is an onnx model?

File: olive/model.py [hunk 6]
```diff
@@ -104,6 +107,25 @@ def prepare_session(
         """
         raise NotImplementedError()
 
+    def download_from_azureml(self, azureml_client: AzureMLClient, download_path: Union[Path, str]):
```
Comment (guotuofeng) [inline olive/model.py:110]: >azureml_client [](http://example.com/codeflow?start=36&length=14)

use MLClient?

File: olive/workflows/run/run.py [hunk 7]
```diff
@@ -124,6 +124,10 @@ def run(config: Union[str, Path, dict], setup: bool = False):
     # input model
     input_model = config.input_model.create_model()
 
+    # Azure ML Client
+    if config.azureml_client:
+        config.engine.azureml_client = config.azureml_client
```
[the remark this answers is not in the document]
Comment (guotuofeng) [reply olive/workflows/run/run.py:129]: shouldn't the config happens in config json?

File: olive/model.py (hunk 6 above)
Comment (xiaoyu-work) [reply olive/model.py:110]: AzureMLClient is a wrapper here for MLClient. Otherwise if we specify MLClient type hint we will need to import MLClient to model.py

File: olive/engine/engine.py [hunk 8]
```diff
@@ -55,6 +57,7 @@ def __init__(
         host: Optional[OliveSystem] = None,
         target: Optional[OliveSystem] = None,
         evaluator_config: Optional[OliveEvaluatorConfig] = None,
+        azureml_client: Optional[AzureMLClientConfig] = None,
```
Comment (guotuofeng) [inline olive/engine/engine.py:60]: >azureml_client [](http://example.com/codeflow?start=8&length=14)

seems useless?

Comment (xiaoyu-work) [reply olive/engine/engine.py:60]: Oh yeah good catch! Will remove this.

Comment (xiaoyu-work) [reply olive/engine/engine.py:60]: Yeah good catch! Will remove this.

File: olive/model.py (hunk 6 above)
Comment (guotuofeng) [reply olive/model.py:110]: We could use "MLClient" as the type hints.

Comment (xiaoyu-work) [reply olive/model.py:110]: Yeah that's what I mean. MLClient is from azure ai package "from azure.ai.ml import MLClient", we don't want to import this in model module level.

File: olive/workflows/run/run.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (guotuofeng) [reply olive/workflows/run/run.py:129]: could not the following works?

"azureml_client": {
    "subscription_id": "<place_holder>",
    "resource_group": "<place_holder>",
    "workspace_name": "<place_holder>"
},
"engine": {
        "search_strategy": {
            "execution_order": "joint",
            "search_algorithm": "random",
            "search_algorithm_config": {
                "num_samples": 3,
                "seed": 0
            }
        },
        "evaluator": "common_evaluator",
        "host": "local_system",
        "target": "local_system",
        "azureml_client": "azureml_client"
        "clean_cache": true,
        "cache_dir": "cache"
    }

File: olive/model.py (hunk 6 above)
Comment (guotuofeng) [reply olive/model.py:110]: we need not import the MLClient. We could either use string or just leave the type hints empty here

File: olive/workflows/run/run.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (guotuofeng) [reply olive/workflows/run/run.py:129]: Please consider that case that there are multiple workspaces in one olive run in future since we all not sure the assumption that only use one workspace is right.

[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply olive/workflows/run/run.py:129]: I read that aml documentation. So the question here is, will Olive handle the sharing operation? Like if the user wants to use the model registered in another workspace, should they complete the sharing operation already?

File: olive/model.py (hunk 6 above)
Comment (xiaoyu-work) [reply olive/model.py:110]: I don't like a placeholder string or empty type hint here. Any reason that model download should not know AzureMLClient?

Comment (guotuofeng) [reply olive/model.py:110]: Because model will have knowledge about the implementation detail of AzureMLClient. This is a design flaw we need avoid.

File: olive/workflows/run/run.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (guotuofeng) [reply olive/workflows/run/run.py:129]: not just about share model. Please consider the case: model is registered in workspace A. but compute target is in workspace B. 

Since no one know whether our current assumption is correct or not, we'd better to keep that in mind when we design/implement to consider the extensibility.

[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply olive/workflows/run/run.py:129]: I just realize that the assets sharing is between workspace and registry but not workspace and workspace. So in this case there will only be one workspace in total.

[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply olive/workflows/run/run.py:129]: We can add support registry support into AMLClient in the future.

File: olive/model.py (hunk 6 above)
Comment (guotuofeng) [reply olive/model.py]: [TYPE_CHECKING](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING) should be the thing your are looking for.

File: olive/engine/engine.py [hunk 9]
```diff
@@ -668,9 +675,19 @@ def _run_passes(
...
                 and not self.host_for_pass(pass_id).system_type == SystemType.AzureML
             ):
-                error_msg = "Azure ML model only supports AzureMLSystem for Olive Pass"
-                logger.error(error_msg)
-                raise Exception(error_msg)
+                if not self.azureml_client_config:
+                    raise ValueError("AzureML client config is required to download the model from AzureML storage")
+                model_download_path = self._model_cache_path / "azureml_input_model"
+                model_path = model.download_from_azureml(
+                    self.azureml_client_config.create_client(), model_download_path
```
Comment (guotuofeng) [inline olive/engine/engine.py:682]: better to move out of the for loop?

Comment (xiaoyu-work) [reply olive/engine/engine.py:682]: Since each pass and evaluator can have its own host and target, I was thinking one case that all previous passes and evaluations are running on AML but one of the middle are running locally. This is to handle that case so I put this in the loop.

**Review by guotuofeng — DISMISSED** [review]

File: olive/engine/engine.py (hunk 9 above)
Comment (xiaoyu-work) [reply olive/engine/engine.py:682]: Although currently, we will need to downlaod the model from the AML after the first pass run on AML, this is a gap from AML side but not our side. We will support this logic in the future once AML fixes this.

File: olive/model.py [hunk 10]
```diff
@@ -10,14 +10,16 @@
...
-from typing import Any, Callable, Dict, List, Optional, Union
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional, Union
 
 import onnx
 import torch
 import yaml
 from onnx import AttributeProto, GraphProto
 from pydantic import validator
 
+if TYPE_CHECKING:
```
Comment (trajepl) [inline olive/model.py:21]: Nice~!

File: olive/systems/azureml/aml_system.py [hunk 11]
```diff
@@ -33,17 +33,17 @@ class AzureMLSystem(OliveSystem):
 
     def __init__(
         self,
-        aml_config_path: str,
+        azureml_client_config: AzureMLClientConfig,
```
Comment (trajepl) [inline olive/systems/azureml/aml_system.py:36]: Should we also keep the legacy `aml_config_path` before the releasement?

**Review by trajepl — DISMISSED** [review]

File: olive/systems/azureml/aml_system.py (hunk 11 above)
Comment (xiaoyu-work) [reply olive/systems/azureml/aml_system.py:36]: Even we keep this, aml system wont't work with aml_config_path as we changed the mlclient creation logic inside. Do we want to keep this?

**Review by guotuofeng — DISMISSED** [review]

**Review by trajepl — APPROVED** [review]
